python_benchmarks/core.py

```diff
@@ -0,0 +1,202 @@
+        self.current_time = 0.0
+
+    def __call__(self, function_to_benchmark: Callable, *args, **kwargs):
+        return self.benchmark(function_to_benchmark, *args, **kwargs)
+
+    def __getattr__(self, attr):
+        if attr not in self.__dict__:
+            return getattr(self.benchmark, attr)
+        return super().__getattr__(attr)
+
+    def torchprofile_timer(self) -> float:
+        """
+        Custom torchprofiler-based timer used by pytest-benchmark.
+        At every timer call, the profiler is stopped to compute the elapsed CUDA time
+        and the global clock is incremented. The profiler is restarted before returning to continue tracing.
+
+        Returns:
+            self.current_time: Global monotonic clock variable
+        """
+        try:
+            self.prof.stop()
+            prof_averages = self.prof.key_averages()
+            elapsed_cuda_time = self._get_kernel_time(prof_averages)
+            self._increment_global_time(elapsed_cuda_time)
+            self.prof.start()
+        except:
+            self.prof.start()
+        return self.current_time
+
+    def _get_kernel_time(self, prof_averages) -> float:
+        """
+        Returns:
+            time_value: Elapsed CUDA time in seconds.
+        """
+        elapsed_cuda_time = sum([event.self_cuda_time_total
+            for event in prof_averages 
+            if event.device_type == DeviceType.CUDA])/ 1e6
+        return elapsed_cuda_time
+
+    def _increment_global_time(self, elapsed_time: float) -> None:
+        self.current_time += elapsed_time
+
+    def cleanup(self) -> None:
+        """
+        Stops a running torchprofiler instance if found.
+        """
+        try:
+            self.prof.stop()
+        except:
+            pass
+
+    def compute_metrics(
+        self, inputs: Union[torch.Tensor, List], outputs: Union[torch.Tensor, List]
+    ) -> None:
+        """
+        Utility function to compute metrics for the target function.
+        Current metrics:
+            IOBytes: Total bytes in inputs + outputs
+            BytesPerSecond: IOBytes * total_rounds / total_time
+            Bandwdith (GBps): BytesPerSecond / (1024**3)
+            % Peak Bandwidth (SOL): 100 * Bandwidth /PEAK_BANDWIDTH
+        """
+        if isinstance(inputs, torch.Tensor):
+            inputs = [inputs]
+        if isinstance(outputs, torch.Tensor):
+            outputs = [outputs]
+
+        iobytes = 0
+        for inp in inputs:
+            if isinstance(inp, torch.Tensor):
+                iobytes += inp.element_size() * inp.numel()
+        for out in outputs:
+            if isinstance(out, torch.Tensor):
+                iobytes += out.element_size() * out.numel()
+
+        self.benchmark.extra_info["IOBytes"] = iobytes
+        bytes_per_second = (
+            iobytes * self.benchmark.stats["rounds"]
+        ) / self.benchmark.stats["total"]
+        self.benchmark.extra_info["BytesPerSecond"] = bytes_per_second
```

Comment:
We are using `BytesPerSecond`/`bytes_per_second` and `Bandwidth` inconsistently, causing unnecessary confusion. Can we pick one and converge? For example, I'd rename `bytes_per_second` to `bandwidth_bps` and `BytesPerSecond` to `Bandwidth (Bps)`.